q2_metadata/normalization/_norm_rules.py

```diff
@@ -0,0 +1,322 @@
+        Apply the rules of checking that the passed column in the correct format.
+        [THIS COMMAND WOULD EITHER INFER THE DTYPE (as of now) OR USE THE
+        RESULT OF AN INITIAL DTYPE GETTER FUNCTION (see _dtypes.py)]
+                    'format:
+                      - bool
+                      - float
+                      - int
+                      - str'
+    expected : list
+        Apply the rules of allowing only expected values.
+        These do allow the missing and blanks.
+                    'expected:
+                        - str
+                        - list'
+    ontology : str
+        Name of the ontology to lookup within for control vocabulary, e.g.
+                    'ontology: Gazetteer'
+    blank : str
+        Apply the rules of checking the allowed blank values, by replacing
+        the not-expected values and numpy's nan (empty) by the passed
+        blank value term, e.g.
+                    'blank:
+                        - not applicable
+                        - not collected
+                        - not provided
+                        - restricted access'
+
+    Raises
+    ------
+    ValueError
+        # If the remote argument is not of ``bool`` or ``str`` type.
+        # If none of the samples in the ordination matrix are in the metadata.
+        # If the data is one-dimensional.
+    KeyError
+        # If there's samples in the ordination matrix but not in the metadata.
+    """
+    def __init__(self, variable_rules_fp: str):
+        """
+        Initialize the class instance for one set of rules beloning
+        to one single metadata variable with it possible attribute, as
+        a rate of one attribute per possible rule, plus the current
+        variable name.
+
+        Parameters
+        ----------
+        variable_rules_fp : str
+            Path to one variable's rules yaml file.
+        """
+        self.variable = basename(variable_rules_fp).split('.yml')[0]
+        self.rules = {
+            'edits': {
+                'remap': {},
+                'normalization': {},
+                'validation': {}
+            },
+            'lookups': {
+                'ontology': None,
+                'expected': []
+            },
+            'allowed': {
+                'blank': None,
+                'missing': None
+            },
+            'format': None
+        }
+        self._process_rule(variable_rules_fp)
+
+    def _process_rule(self, variable_rules_fp: str):
+        """Read the yaml file of the current variable's rules.
+
+        Parameters
+        ----------
+        variable_rules_fp : str
+            Path to one variable's rules yaml file.
+
+        Returns
+        -------
+
+        """
+        with open(variable_rules_fp) as handle:
+            variable_rules = yaml.load(handle, Loader=yaml.FullLoader)
+            self._check_variable_rules(variable_rules)
+
+    def _check_variable_rules(self, variable_rules: dict):
```

Comment:
this is extremely complex and almost certainly has a massive cyclomatic complexity score -- it needs to be decomposed.

Reply:
I would need your advice but I guess I could start by making the collection of the parsed information and then apply the checks on this parsed info in another operation of the __init__, instead of having the checks done inside the `with open(variable_rules_fp) as handle`... does that sound less cyclomatically complex?

Notably, I was thinking of making user-defined exceptions classes (see last answer to last question from Yoshiki above). Indeed, the content of method `_check_variable_rules()` is per-rule and hence there could be one Exception class per rule (inherited from a base Error class). What do you think?

Thanks for the quick review!

Reply:
I think I have addressed this by making this `_check_variable_rules()` function to run not inside the read loop but at the end of the `__init__` of the `Rules()` class. Is it what I should do to reduce the cyclomatic complexity score?

For more info on the changes that accompany this, please see:
https://github.com/ElDeveloper/q2-metadata/pull/3?_pjax=%23js-repo-pjax-container#discussion_r401995524

Reply:
I think I have reduced this cyclomatic complexity score by separating the sequence of event as described in https://github.com/ElDeveloper/q2-metadata/pull/3#discussion_r401995524. 

The checking of the rules formatting is made in the `__init__()` of the `Rules()` class, when this class is instantiated, for each variable. I am not sure I address totally the problem and any advice is welcome.